contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,773 @@
+
+    uint refundInCoverAsset = 0;
+
+    if (!lastCoverSegment.expired) {
+      uint32 remainingPeriod = lastCoverSegment.start + lastCoverSegment.period - uint32(block.timestamp);
+
+      PoolAllocation[] storage originalPoolAllocations = coverSegmentAllocations[coverId][lastCoverSegmentIndex];
+
+      {
+        uint originalPoolAllocationsCount = originalPoolAllocations.length;
+
+        // rollback previous cover
+        for (uint i = 0; i < originalPoolAllocationsCount; i++) {
+          stakingPool(originalPoolAllocations[i].poolId).deallocateStake(
+            cover.productId,
+            lastCoverSegment.start,
+            lastCoverSegment.period,
+            originalPoolAllocations[i].coverAmountInNXM,
+            originalPoolAllocations[i].premiumInNXM,
+            lastCoverSegment.globalRewardsRatio
+          );
+
+          originalPoolAllocations[i].premiumInNXM *= (lastCoverSegment.period - remainingPeriod) / lastCoverSegment.period;
+
+          // compute NXM rewards deallocated
+          uint deallocatedRewardsInNXM = originalPoolAllocations[i].premiumInNXM
+          * remainingPeriod / lastCoverSegment.period
+          * lastCoverSegment.globalRewardsRatio / REWARD_DENOMINATOR;
+          tokenController().burnStakingPoolNXMRewards(deallocatedRewardsInNXM, originalPoolAllocations[i].poolId);
+        }
+      }
+
+      refundInCoverAsset = lastCoverSegment.priceRatio
+        * lastCoverSegment.amount
+        / PRICE_DENOMINATOR
+        * remainingPeriod
+        / MAX_COVER_PERIOD;
+
+      // edit cover so it ends at the current block
+      lastCoverSegment.period = lastCoverSegment.period - remainingPeriod;
+    }
+
+    uint totalPremiumInNXM = _buyCover(buyCoverParams, coverId, poolAllocations);
+
+    handlePaymentAndRefund(buyCoverParams, totalPremiumInNXM, refundInCoverAsset);
+
+    // update total cover amount for asset by decreasing the previous amount and adding the new amount
+    if (coverAmountTrackingEnabled) {
+       totalActiveCoverInAsset[cover.payoutAsset] =
+        totalActiveCoverInAsset[cover.payoutAsset] - lastCoverSegment.amount + buyCoverParams.amount;
+    }
+
+    emit CoverEdited(coverId, cover.productId, lastCoverSegmentIndex + 1, msg.sender);
+  }
+
+  function handlePaymentAndRefund(
+    BuyCoverParams memory buyCoverParams,
+    uint totalPremiumInNXM,
+    uint refundInCoverAsset
+  ) internal {
+
+    IPool _pool = pool();
+
+    // calculate refundValue in NXM
+    uint refundInNXM = refundInCoverAsset * 1e18 / _pool.getTokenPrice(buyCoverParams.payoutAsset);
+
+    if (refundInNXM >= totalPremiumInNXM) {
+      // no extra charge for the user
+      return;
+    }
+
+    uint tokenPriceInPaymentAsset = _pool.getTokenPrice(buyCoverParams.paymentAsset);
+    (, uint8 paymentAssetDecimals) = _pool.coverAssets(buyCoverParams.paymentAsset);
+
+    uint premiumInPaymentAsset = totalPremiumInNXM * (tokenPriceInPaymentAsset / 10 ** paymentAssetDecimals);
```

Comment:
The conversion from nxm to asset uses `paymentAssetDecimals` here, however the same conversion uses `1e18` on line 264 (the first parameter to the pushed cover segment `_coverSegments[coverId].push(CoverSegment(...)`)